src/plugins/controls/server/control_types/options_list/options_list_queries.ts

```diff
@@ -0,0 +1,174 @@
+      [key: string]: { doc_count: number };
+    };
+    return rawInvalidSuggestions && !isEmpty(rawInvalidSuggestions)
+      ? Object.entries(rawInvalidSuggestions)
+          ?.filter(([, value]) => value?.doc_count === 0)
+          ?.map(([key]) => key)
+      : [];
+  },
+});
+
+/**
+ * Suggestion aggregations
+ */
+export const getSuggestionAggregationBuilder = ({
+  fieldSpec,
+  textFieldName,
+  searchString,
+}: OptionsListRequestBody) => {
+  if (textFieldName && fieldSpec?.aggregatable && searchString) {
+    return suggestionAggSubtypes.keywordAndText;
+  }
+  if (fieldSpec?.type === 'boolean') {
+    return suggestionAggSubtypes.boolean;
+  }
+  if (fieldSpec && getFieldSubtypeNested(fieldSpec)) {
+    return suggestionAggSubtypes.subtypeNested;
+  }
+  return suggestionAggSubtypes.keywordOnly;
+};
+
+const getEscapedQuery = (q: string = '') =>
+  q.replace(/[.?+*|{}[\]()"\\#@&<>~]/g, (match) => `\\${match}`);
+
+const suggestionAggSubtypes: { [key: string]: OptionsListAggregationBuilder } = {
+  /**
+   * the "Keyword only" query / parser should be used when the options list is built on a field which has only keyword mappings.
+   */
+  keywordOnly: {
+    buildAggregation: ({ fieldName, searchString }: OptionsListRequestBody) => ({
+      terms: {
+        field: fieldName,
+        include: `${getEscapedQuery(searchString)}.*`,
+        execution_hint: 'map',
+        shard_size: 10,
+      },
+    }),
+    parse: (rawEsResult) =>
+      get(rawEsResult, 'aggregations.suggestions.buckets')?.map(
+        (suggestion: { key: string }) => suggestion.key
+      ),
+  },
+
+  /**
+   * the "Keyword and text" query / parser should be used when the options list is built on a multi-field which has both keyword and text mappings. It supports case-insensitive searching
+   */
+  keywordAndText: {
+    buildAggregation: (req: OptionsListRequestBody) => {
+      if (!req.textFieldName) {
+        // if there is no textFieldName specified, or if there is no search string yet fall back to keywordOnly
+        return suggestionAggSubtypes.keywordOnly.buildAggregation(req);
+      }
+      const { fieldName, searchString, textFieldName } = req;
+      return {
+        filter: {
+          match_phrase_prefix: {
+            [textFieldName]: getEscapedQuery(searchString),
+          },
+        },
+        aggs: {
+          keywordSuggestions: {
+            terms: {
+              field: fieldName,
+              shard_size: 10,
+            },
+          },
+        },
+      };
+    },
+    parse: (rawEsResult) =>
+      get(rawEsResult, 'aggregations.suggestions.keywordSuggestions.buckets')?.map(
+        (suggestion: { key: string }) => suggestion.key
+      ),
+  },
+
+  /**
+   * the "Boolean" query / parser should be used when the options list is built on a field of type boolean. The query is slightly different than a keyword query.
+   */
+  boolean: {
```

Comment:
Hmmm... Is this necessary? Because `boolean` fields don't have a search box when selecting, and `getSuggestionAggregationBuilder` is only used there, right? Unless I misunderstood something :) 

![image](https://user-images.githubusercontent.com/8698078/166968807-10a0c372-1d13-4bb8-9655-791ac2141c5c.png)

Reply:
`getSuggestionAggregationBuilder` is actually used for getting the suggestions in the query context. Even for Boolean fields we actually query the ES index rather than just returning a static `['true', 'false']`.

 This is for validation / chaining purposes. If you had one control that said `Cancelled is false` then you had another `Cancelled` control after it, the second control would only show `false` because we actually run the `suggestion` aggregation.

Now I'm doubting the naming of this aggregation builder too bahaha, because it's more generic than just **suggestions** 🕵️